frontend/src/modules/editor/actions.js

```diff
@@ -44,6 +45,30 @@ export function updateSelection(content: string): UpdateSelectionAction {
 }
 
 
+/**
+ * Update failed checks in the active editor.
+ */
+export type FailedChecks = {|
+    +clErrors: Array<string>,
+    +pErrors: Array<string>,
+    +clWarnings: Array<string>,
+    +pndbWarnings: Array<string>,
+    +ttWarnings: Array<string>,
```

Comment:
I know, I know... I looked into refactoring checks before digging into it, but the patch would be pretty big. Maybe I should file a P5 Translate.Next bug, so we don't forget about it?

I'm thinking checks should instead be formatted like this:
```js
failedChecks = {
    errors: [
        {
            message: 'Missing variable %s',
            library: 'compare-locales'
        },
        ...
    ],
    warnings: [...],
}
```

Reply:
Oh god yes that looks so much better! What I was wondering is if you could maybe add a layer between the API and the reducer to make the data from the server look more like what you show up there? It might be too much work, though, feel free not to do that.

However, yes, please do file a bug about refactoring errors and warnings. :-) Not a Translate.Next though, I think it's a general Pontoon issue.

Reply:
I'm thinking the bug I'm about to file would make such API redundant, so I'd rather not spent time working on it.

Reply:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1545740.

@jotes

Reply:
@mathjazz thanks! I totally agree.  The current structure of the API response is a bit complicated (I didn't consider the later use and scaling). Your proposal looks very good :-)